### ISSUE TITLE
Use with Celluloid & ActiveRecord

### ISSUE THREAD

**Issue by ioquatix** [brianmario/mysql2#658]
Is it possible to use this with Celluloid and ActiveRecord, e.g. in a `Celluloid::IO` reactor so that queries do not block?

I don't mind helping but it would be good to know if there is any effort towards this goal.

**Comment by sodabrew** [comment]
This should be possible – there's an EventMachine wrapper here: https://github.com/brianmario/mysql2/blob/master/lib/mysql2/em.rb – I am not familiar with Celluloid, but I am happy to help and would merge a reasonable PR to support Celluloid with mysql2.

**Comment by ioquatix** [comment]
The Celluloid::IO event reactor could work if you can wrap read and write calls to the socket.. is it possible to change the class of the socket used for connections or is this hardcoded by the mysql library?

**Comment by sodabrew** [comment]
Unfortunately the read and write calls are deep in the MySQL library. It does provide a hook to get the file descriptor number, which the EM driver uses to watch for activity.

**Comment by ioquatix** [comment]
So, if you only have a single thread.. would that break ActiveRecord connection pooling? Like, if you have a single thread doing, say, 10 in-flight queries?

**Comment by sodabrew** [comment]
The MySQL protocol can only handle a single in-flight query per connection, so the main job of the connection pool is to keep track of which connections are in-use vs. available for a query. A single thread could send queries on multiple connections; the query write will block until written, then return, and the fd can be checked with `select()` to see if it's ready to be read. At that point the result reads will block until they are done reading. It's not perfect, one large result body could leave some smaller results waiting for a while.

**Comment by ioquatix** [comment]
Is it possible for a single thread to have more than one active connection? Does this integrate well with ActiveRecord?

**Comment by ioquatix** [comment]
My experience with ActiveRecord is that each thread checks out a unique connection, does some process, then returns the connection.

The problem would be, is it possible for a single thread to check out another connection if the first one is blocked?

**Comment by sodabrew** [comment]
Oh, interesting detail. I'm not sure. Would be worth a look to see how AR handles this.

**Comment by ioquatix** [comment]
My understanding is that AR checks out connections per-thread. It uses thread local storage to lookup the current connection. I may be wrong on this. But, if you do `User.find(x)` it would use the thread's current connection.

The current approach for handling this is to have a separate thread/actor for handling mysql queries. It works well but isn't intuitive. ActiveRecord simply might not be suitable to work multiple asynchronous connections on the same thread.

**Comment by ioquatix** [comment]
Is it possible that the same thread enters this block multiple times? https://github.com/brianmario/mysql2/blob/3907e94b1311d6cccd2a58004b15f570404ca728/lib/mysql2/em.rb#L44-L50

If so, how does the actual socket multiplex the requests?

**Comment by sodabrew** [comment]
The socket does not multiplex requests. Each connection to MySQL is a stateful connection, and can be waiting, sending a query, receiving a result, in the middle of a transaction, etc.

**Comment by TiagoCardoso1983** [comment]
Hi all, I'm also working on something like this. @ioquatix , I don't think you can wrap the socket in a celluloid-enhanced class, as this happens within the C extension. However, the file descriptor can suffice.  @sodabrew , is this socket parameter a Ruby Socket? If so, one could write a similar extension:

```ruby
def query(sql, opts={})
    if ::Celluloid::IO::evented?
      ::Celluloid::IO.wait_writable(self.socket)
      super(sql, opts.merge(:async => true))
      ::Celluloid::IO.wait_readable(self.socket)
      result = @client.async_result
   else
      super(sql, opts)
   end
end
```

What do you guys think? This was also the subject of a celluloid pg-listen example gist on how to interact with a socket exposed by a database library. One could use something similar. Also, I don't think that the multiplex is the most relevant scenario here, important would be that mysql queries wouldn't block the Celluloid::IO Actor.

**Comment by ioquatix** [comment]
In the end, what I did was write a pool of actors which provides a high level interface for database requests. ActiveRecord is configured with N+K connections in the connection pool and I allocate N database actors in the pool. This has worked really well and isolated database issues to the database actor.

**Comment by TiagoCardoso1983** [comment]
I think, with this patch, you could have 1 actor handling your pool of connections, less thread overhead and better resource utilisation, what do you think?

**Comment by ioquatix** [comment]
Perhaps, it would be nice to try it out in a simulated production environment.

**Comment by ioquatix** [comment]
Also, ActiveRecord only has one connection per thread. I didn't want to break this.

**Comment by sodabrew** [comment]
@TiagoCardoso1983 The MySQL fd is either a Unix Domain socket or a TCP socket, both at the system level and not wrapped by any Ruby objects. The sockets operate in a blocking mode. It looks like MariaDB added a whole new set of API calls in order to implement non-blocking behavior (mysql2 doesn't support these, I just read about it forty seconds ago for the first time) https://mariadb.com/kb/en/mariadb/using-the-non-blocking-library/

The thread pool / connection pool model is the only reasonable one I can think of. Maybe a Celluloid wrapper class around the entire Mysql2::Client class would make sense, so that it provides the mailbox methods and state representation that can tie into the Celluloid::Actor or Celluloid::Pool interface? (Am I reading https://github.com/celluloid/celluloid-pool right?)

**Comment by TiagoCardoso1983** [comment]
@sodabrew MariaDB might be an improvement, nevertheless Cellulod::IO must get away with a similar solution as the one created for EventMachine, wish means that it'll be non-blockish. Non necessarily to do with the pool. I'll try to come up with an (hopefully) working example.

**Comment by TiagoCardoso1983** [comment]
So, I finally came around it. https://gist.github.com/TiagoCardoso1983/2c44a44fc6858e6f0b58

**Comment by TiagoCardoso1983** [comment]
@ioquatix , I'm of the opinion that maybe Celluloid::IO should provide a better API to abstract raw socket waits. Unfortunately one can not always use the proxy Socket classes or pass them around and need a more bare-bones approach, as most of the network clients (not only mysql2) are written as C extensions and not all of them give back the descriptor ID as mysql2 does.

**Comment by thedrow** [comment]
Is there a pull request for this?

**Comment by TiagoCardoso1983** [comment]
@thedrow no, as I didn't have the time to write tests and I don't have a project using mysql to test the robustness of the wrapper client. Did you try it on your side? Is it working out for you?

**Comment by thedrow** [comment]
I am not using Celluloid yet but I'm considering it.

**Comment by TiagoCardoso1983** [comment]
So apply the patch and see how it works on your side. As a side note, you might be careful when handling X database connections in an event loop and this X being unbounded. Most non-NOJS databases handle unbounded number of connections very bad, and you'd probably best use the suggestion from @ioquatix , which is forget celluloid-io, used celluloid-pool instead, and have a db connection per actor.

**Comment by ioquatix** [comment]
Haha, I just read this thread again. Wow, good times.

**Comment by sodabrew** [comment]
Closing — Celluloid itself has been unmaintained for years.
